features/steps/steps.py

```diff
@@ -1,33 +1,294 @@
+    return temporary_file_name
+
+def draft_action_run(node_value,input_value):
+    # Helper function to 'draft' a given input_value and write the DOI to a temporary file.
+    # This file will be available for other validation functions.
+    db_name = 'doi_temp.db'
+    _action = DOICoreActionDraft(db_name=db_name)
+    logger.info(f"input_value {input_value}")
+
+    o_doi_label = _action.run(input=input_value,
+                          node=node_value,
+                          submitter='my_user@my_node.gov',force=True)
+    # Save o_doi_label to disk so it can be compared to historical in next step
+    logger.info(f"success input_value {input_value}")
+    return save_doi_to_temporary_file(o_doi_label)
+
+def reserve_action_run(node_value,input_value):
+    # Helper function to 'reserve' a given input_value.
+    global g_action_expectant_bad
+    logger.info(f"when node_value,input_value {node_value,input_value}")
+    o_doi_label = None
+
+    db_name = 'doi_temp.db'
+    _action = DOICoreActionReserve(db_name=db_name)
+
+    o_doi_label = _action.run(
+                      input=input_value,
+                      node=node_value, submitter='my_user@my_node.gov',
+                      dry_run=True,force=True)
+
+    return save_doi_to_temporary_file(o_doi_label)
+    #return o_doi_label
+
+def draft_output_compare(output_file, ref_output_value):
+    # Function compare two XML files created from 'draft' or 'reserve' actions.
+    # Assumption(s): 
+    #   1.  The name of the new XML file is defined in get_temporary_output_filename().
+    #   2.  The name of the historical name is ref_output_value
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+
+    o_fields_differ_list, o_values_differ_list, o_record_index_differ_list = DOIDiffer.doi_xml_differ(ref_output_value,
+                                                                                                      output_file)
+
+
+    logger.info(f'different fields are {o_fields_differ_list}')
+
+    assert len(o_fields_differ_list) is 0
+
+    return 1
 
-@given('a valid PDS4 {input_type} at url {input_value}')
-def given_valid_input(context, input_type, input_value):
-    logger.info(f"given {input_type} {input_value}")
+def reserve_output_compare(output_file, ref_output_value):
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+    # Use the same function draft_output_compare() to compare a DOI (XML output) from the 'reserve' action.
+    draft_output_compare(output_file,ref_output_value)
 
+def initialize_normal_draft():
+    global g_action_expectant_bad, g_draft_condition
+    # Set to g_action_expectant_bad to signify to not expect the action to be bad.
+    g_action_expectant_bad = False
+    g_draft_condition = TestCondition.NORMAL
 
-@given('an invalid PDS4 {input_type} at {input_value}')
+def initialize_normal_reserve():
+    global g_action_expectant_bad, g_reserve_condition
```

Comment:
All references to these global variables have been removed and the behave testing runs OK.

Reply:
Good, that is more simple now. I removed the reserve_compare function which is the same as file_compare. I believe also we don't absolutely need to write a local temporary file before comparing the results (I mean that would be more robust if we don't need to write something on a local system), but since that work and everything is much more clear and neat now, I am happy with that ! Thanks